src/Mafia/Init.hs

```diff
@@ -2,193 +2,163 @@
-      -- done because we'll be reinstalling it later if it's required.
-
-      result <- liftIO . runEitherT $ do
-        Hush <- cabal "sandbox" ["hc-pkg", "--", "unregister", "--force", pkg]
-        return ()
-
-      case result of
-        Left  _ -> return ()
-        Right _ -> liftIO (T.putStrLn ("Sandbox: Unregistered " <> pkg))
-
-invalidateCache :: MonadIO m => File -> m ()
-invalidateCache cabalFile = do
-  mtxt <- readUtf8 cabalFile
-  case mtxt of
-    Nothing  -> return ()
-    Just txt -> do
-      time <- liftIO getCurrentTime
-      let header = "-- " <> T.pack (show time) <> "\n"
-      writeUtf8 cabalFile (header <> txt)
-
-cacheUpdate :: (Functor m, MonadIO m) => File -> File -> m (Maybe CacheUpdate)
-cacheUpdate src dst = do
-  src_missing <- not <$> doesFileExist src
-  dst_missing <- not <$> doesFileExist dst
-  if | src_missing && dst_missing -> return (Nothing)
-     | src_missing                -> return (Just (Delete dst))
-     | dst_missing                -> return (Just (Update src dst))
-     | otherwise                  -> cacheUpdateDiff src dst
-
-cacheUpdateDiff :: MonadIO m => File -> File -> m (Maybe CacheUpdate)
-cacheUpdateDiff src dst = do
-  src_bytes <- readBytes src
-  dst_bytes <- readBytes dst
-  if | src_bytes == dst_bytes -> return (Nothing)
-     | otherwise              -> return (Just (Update src dst))
-
-getCacheDir :: EitherT MafiaError IO Directory
-getCacheDir = do
-  cacheDir <- (</> "mafia") <$> liftCabal initSandbox
-  createDirectoryIfMissing False cacheDir
-  return cacheDir
+  firstEitherT InitGitError initSubmodules
+
+  sandboxDir <- firstEitherT InitCabalError initSandbox
+  let statePath = sandboxDir </> "mafia/state.json"
+
+  liftIO (T.putStrLn "Checking for changes to dependencies...")
+  previous <- readMafiaState statePath
+  current  <- getMafiaState
+  hasDist  <- liftIO $ doesDirectoryExist "dist"
+
+  packages <- checkPackages
+
+  when (previous /= Just current || not hasDist || packages == PackagesBroken) $ do
+    liftIO (T.putStrLn "Installing dependencies...")
+    let sdeps = Set.toList (msSourceDependencies current)
+    firstEitherT InitInstallError $ installDependencies sdeps
+    firstEitherT InitCabalError $ cabal_ "configure" ["--enable-tests" , "--enable-benchmarks"]
+    writeMafiaState statePath current
+
+------------------------------------------------------------------------
+
+data PackageState =
+    PackagesOk
+  | PackagesBroken
+    deriving (Eq, Ord, Show)
+
+checkPackages :: EitherT InitError IO PackageState
+checkPackages = do
+  packageDB     <- firstEitherT InitCabalError getPackageDB
+  packages      <- getDirectoryListing (RecursiveDepth 0) packageDB
+  packagesExist <- and <$> traverse doesFileExist packages
+  case packagesExist of
+    True  -> return PackagesOk
+    False -> return PackagesBroken
+
+------------------------------------------------------------------------
+
+data MafiaState =
```

Comment:
Yeah probably, originally it was the only thing in here and was more like `InitState`